src/util/threadpool.rs

```diff
@@ -242,54 +249,63 @@ where
     C: Context,
 {
     fn new(
-        task_queue: Arc<(Mutex<FifoQueue<C>>, Condvar)>,
+        state: Arc<(Mutex<ScheduleState<C>>, Condvar)>,
         task_count: Arc<AtomicUsize>,
         tasks_per_tick: usize,
-        stop_flag: Arc<AtomicBool>,
         ctx: C,
     ) -> Worker<C> {
         Worker {
-            stop_flag: stop_flag,
-            task_queue: task_queue,
+            state: state,
             task_count: task_count,
             tasks_per_tick: tasks_per_tick,
             task_counter: 0,
             ctx: ctx,
         }
     }
 
-    fn get_task_timeout(&mut self, timeout: Option<Duration>) -> Option<Task<C>> {
-        let &(ref lock, ref cvar) = &*self.task_queue;
-        let mut task_queue = lock.lock().unwrap();
-
-        if let Some(task) = task_queue.pop() {
-            return Some(task);
+    fn next_task(&mut self, timeout: Duration) -> Option<Task<C>> {
```

Comment:
I think we do not need to pass timeout, it always be `NAP_SECS` and it does not return after timeout.